docs/DesignDocs/OpteeIntegration.md

```diff
@@ -0,0 +1,321 @@
+have been there to begin with.
+
+After the TA loader lays out the binary in memory and has had the TA
+instance initialize itself, a session handle is returned to the CA. This
+session handle is then used by the CA for subsequent communication with
+the TA instance.
+
+#### Image Validation
+
+When a TA is read from the REE's filesystem, the REE TA store checks that
+the signature of the hash is correct. Then, it computes the hash as per the
+formulas above, choosing which one depending on whether the TA is encrypted,
+and verifies that it matches the signed hash. If everything checks out, the
+cleartext, standard ELF binary is passed to the ELF loader proper.
+
+#### TA Stores & Open Enclave
+
+A TA is a TA regardless of where it is stored. The three TA stores, namely
+the REE TA store, the Secure Storage TA store, and the Early TA store,
+retrieve regular TA binaries. So, if you build a TA with the Open Enclave SDK,
+you can make that into an early TA by feeding it to OP-TEE's build system,
+which will embed the TA binary into OP-TEE itself. Similarly, you can take
+an Open Enclave TA and install it into secure storage.
+
+There is a Pseudo-TA (PTA, a kernel-mode extension of OP-TEE whose interface
+boundaries are those of a regular TA) called the Secure Storage Management
+PTA (henceforth, secstor PTA). The secstor PTA exposes an install command to
+the non-secure world. This commands takes a TA binary that sits in the REE's
+filesystem, performs the same validation steps as the REE TA store on the
+TA binary header (except that encrypted TAs cannot be installed), and if it
+all checks out, the cleartext, standard ELF is written out to secure storage
+(either to the RPMB, or to an encrypted blob on the REE's filesystem for
+which only OP-TEE has the key).
+
+What you can't do currently, however, is use the Open Enclave SDK to write
+PTAs. This might be feasible in the future, but PTAs are meant to easily and
+cleanly extend OP-TEE, not as a general-purpose runtime.
+
+### TA Metadata
+
+The metadata for an enclave is split between the TA binary header, the TA
+header, and global variables. The first was discussed above and said to be
+consumed exclusively by the REE TA store.
+
+The TA header includes the UUID of the TA, the requested stack size, and flags
+that describe the desired runtime behaviour of the TA (see section 2.10.5 in
+the [OP-TEE documentation](https://optee.readthedocs.io/_/downloads/en/latest/pdf/)).
+The TA header is placed in an ELF section named `.ta_head` and is explicitly
+located at the top of the resulting ELF binary via a linker script. The TA
+header is only used at load time.
+
+The primary global variable, `ta_props`, is used by libutee. libutee is the
+secure user-mode library that implements the
+[GlobalPlatform TEE Internal Core API](https://globalplatform.org/specs-library/tee-internal-core-api-specification-v1-2/)
+for OP-TEE. This API is the oeenclave-equivalent for GlobalPlatform, and
```

Comment:
Can you  elaborate a bit more on "This API is the oeenclave-equivalent for GlobalPlatform"?

Reply:
That observation was drawing a parallel between GP APIs and OE APIs: the GP TEE Client API is to the host-side APIs of the OE SDK what the GP TEE Internal Core API is to the enclave-side APIs of the OE SDK.

The programming model for both GP and OE require that the host and enclave have access to a set of APIs so that they both be able to, one, communicate with each other, and, two, carry out their functions (roughly speaking, of course).

If you draw a diagram of the API layers for GP and OE, you get the same thing:

```
    Host                Host
------------ <=> -----------------
OE Host APIs     GP TEE Client API

    Enclave                 Enclave
--------------- <=> ------------------------
OE Enclave APIs     GP TEE Internal Core API
```

The functionality exposed by the GP and OE APIs for both the host and enclave have some overlap, as well as some differences.

That's what my comment was getting at.

Reply:
Added paragraph explaining this